File: android/app/src/main/java/com/example/app/domain/navcaddy/memory/PatternDecayCalculator.kt
package caddypro.domain.navcaddy.memory

import kotlin.math.pow

/**
 * Calculator for time-based pattern confidence decay.
 *
 * Implements exponential decay to prevent stale patterns from dominating advice.
 * Uses a 14-day half-life per spec resolution Q5.
 *
 * Spec reference: navcaddy-engine.md R5, Q5 (14-day decay half-life)
 */
class PatternDecayCalculator(
    private val decayHalfLifeDays: Double = DEFAULT_HALF_LIFE_DAYS
) {
    companion object {
        /** Default decay half-life in days (spec Q5) */
        const val DEFAULT_HALF_LIFE_DAYS = 14.0

        /** Milliseconds per day for time conversions */
        private const val MILLIS_PER_DAY = 24 * 60 * 60 * 1000L

        /** Maximum age in days before confidence reaches near-zero (6 half-lives = ~1.5%) */
        private const val MAX_AGE_DAYS = 84.0 // 6 half-lives
    }

    /**
     * Calculate the decay factor for a shot based on time elapsed.
     *
     * Uses exponential decay formula: decay = 0.5^(age / half_life)
     *
     * @param shotTimestamp Timestamp when the shot was recorded (milliseconds)
     * @param currentTime Current timestamp (milliseconds)
     * @return Decay factor between 0.0 (very old) and 1.0 (recent)
     */
    fun calculateDecay(shotTimestamp: Long, currentTime: Long = System.currentTimeMillis()): Float {
        require(shotTimestamp <= currentTime) {
            "Shot timestamp ($shotTimestamp) cannot be in the future"
        }

        val ageMillis = currentTime - shotTimestamp
        val ageDays = ageMillis.toDouble() / MILLIS_PER_DAY

        // If age exceeds max, return minimum decay
        if (ageDays >= MAX_AGE_DAYS) {
            return 0.0f
        }

        // Exponential decay: 0.5^(age / half_life)
        val decay = 0.5.pow(ageDays / decayHalfLifeDays)
        return decay.toFloat()
    }

    /**
     * Calculate decayed confidence for a pattern.
     *
     * Applies decay to the base confidence to reduce influence of old patterns.
     *
     * @param baseConfidence Original confidence value (0-1)
     * @param lastOccurrence Timestamp of most recent occurrence (milliseconds)
     * @param currentTime Current timestamp (milliseconds)
     * @return Decayed confidence between 0.0 and 1.0
     */
    fun calculateDecayedConfidence(
        baseConfidence: Float,
        lastOccurrence: Long,
        currentTime: Long = System.currentTimeMillis()
    ): Float {
        require(baseConfidence in 0f..1f) {
            "Base confidence must be between 0 and 1, was: $baseConfidence"
        }

        val decayFactor = calculateDecay(lastOccurrence, currentTime)
        return baseConfidence * decayFactor
    }

    /**
     * Calculate the age in days for a timestamp.
     *
     * @param timestamp Timestamp to calculate age for (milliseconds)
     * @param currentTime Current timestamp (milliseconds)
     * @return Age in days
     */
    fun calculateAgeDays(timestamp: Long, currentTime: Long = System.currentTimeMillis()): Double {
        val ageMillis = currentTime - timestamp
        return ageMillis.toDouble() / MILLIS_PER_DAY
    }

    /**
     * Check if a timestamp is within the retention window.
     *
     * @param timestamp Timestamp to check (milliseconds)
     * @param retentionDays Retention window in days (default 90 per Q5)
     * @param currentTime Current timestamp (milliseconds)
     * @return True if within retention window
     */
    fun isWithinRetentionWindow(
        timestamp: Long,
        retentionDays: Int = 90,
        currentTime: Long = System.currentTimeMillis()
    ): Boolean {
        val ageDays = calculateAgeDays(timestamp, currentTime)
        return ageDays < retentionDays
    }
}

File: ios/App/Sources/Core/Domain/NavCaddy/Memory/PatternDecayCalculator.swift
import Foundation

/// Calculates time-based decay for miss pattern confidence.
///
/// Spec R5: Applies time-based decay to older patterns to avoid permanent anchoring.
/// Q5: 14-day decay half-life.
struct PatternDecayCalculator {
    /// Decay half-life in days (default: 14 days per spec Q5)
    let decayHalfLifeDays: Double

    /// Decay function type
    enum DecayFunction {
        case exponential
        case linear
    }

    let decayFunction: DecayFunction

    init(
        decayHalfLifeDays: Double = 14.0,
        decayFunction: DecayFunction = .exponential
    ) {
        self.decayHalfLifeDays = max(decayHalfLifeDays, 1.0) // Minimum 1 day
        self.decayFunction = decayFunction
    }

    /// Calculates decay factor for a given timestamp.
    ///
    /// - Parameters:
    ///   - timestamp: The timestamp of the original data point
    ///   - currentDate: The current date for decay calculation (default: now)
    /// - Returns: Decay factor between 0.0 (very old) and 1.0 (recent)
    func calculateDecay(for timestamp: Date, at currentDate: Date = Date()) -> Double {
        let daysSinceTimestamp = currentDate.timeIntervalSince(timestamp) / (24 * 60 * 60)

        // Don't decay future timestamps or negative time intervals
        guard daysSinceTimestamp >= 0 else {
            return 1.0
        }

        switch decayFunction {
        case .exponential:
            return exponentialDecay(days: daysSinceTimestamp)
        case .linear:
            return linearDecay(days: daysSinceTimestamp)
        }
    }

    // MARK: - Private Decay Functions

    /// Exponential decay: decay_factor = 0.5^(days / half_life)
    private func exponentialDecay(days: Double) -> Double {
        return pow(0.5, days / decayHalfLifeDays)
    }

    /// Linear decay: decay_factor = max(0, 1 - (days / (2 * half_life)))
    /// Reaches 0 at 2 * half_life
    private func linearDecay(days: Double) -> Double {
        let decayWindow = 2.0 * decayHalfLifeDays
        return max(0.0, 1.0 - (days / decayWindow))
    }
}

// MARK: - Convenience Extensions

extension PatternDecayCalculator {
    /// Default calculator with 14-day exponential decay (spec Q5)
    static let `default` = PatternDecayCalculator()

    /// Short-term calculator with 7-day decay for recent patterns
    static let shortTerm = PatternDecayCalculator(decayHalfLifeDays: 7.0)

    /// Long-term calculator with 30-day decay for persistent patterns
    static let longTerm = PatternDecayCalculator(decayHalfLifeDays: 30.0)
}
